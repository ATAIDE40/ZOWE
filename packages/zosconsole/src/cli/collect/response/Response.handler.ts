/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { ICommandHandler, IHandlerParameters, Session } from "@brightside/imperative";
import { CollectCommand, ICollectParms, IConsoleResponse } from "../../../../../zosconsole";

/**
 * Handle to collect a MVS console command response
 * @export
 * @class Handler
 * @implements {ICommandHandler}
 */
export default class Handler implements ICommandHandler {

    public async process(commandParameters: IHandlerParameters) {

        const profile = commandParameters.profiles.get("zosmf");
        const session = new Session({
            type: "basic",
            hostname: profile.host,
            port: profile.port,
            user: profile.user,
            password: profile.pass,
            base64EncodedAuth: profile.auth,
            rejectUnauthorized: profile.rejectUnauthorized,
        });
        const collectParms: ICollectParms = {
            commandResponseKey: commandParameters.arguments.responsekey,
            consoleName: commandParameters.arguments["console-name"],
        };

        const response: IConsoleResponse = await CollectCommand.collect(session, collectParms);

        // Print out the response
        commandParameters.response.console.log(response.commandResponse);

        // Return as an object when using --response-format-json
        commandParameters.response.data.setObj(response);

    }
}
