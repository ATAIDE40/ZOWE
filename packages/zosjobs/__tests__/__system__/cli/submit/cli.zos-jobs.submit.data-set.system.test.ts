/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { ITestEnvironment } from "./../../../../../../__tests__/__src__/environment/doc/response/ITestEnvironment";
import { TestEnvironment } from "../../../../../../__tests__/__src__/environment/TestEnvironment";
import { runCliScript } from "./../../../../../../__tests__/__src__/TestUtils";
import { TestProperties } from "../../../../../../__tests__/__src__/properties/TestProperties";
import { ITestSystemSchema } from "../../../../../../__tests__/__src__/properties/ITestSystemSchema";

process.env.FORCE_COLOR = "0";

// Test Environment populated in the beforeAll();
let TEST_ENVIRONMENT: ITestEnvironment;

let account: string;
let defaultSystem: ITestSystemSchema;
let systemProps: TestProperties;
let jclMember: string;
let psJclDataSet: string;
describe("zos-jobs submit data-set command", () => {
    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            testName: "zos_jobs_submit_command",
            tempProfileTypes: ["zosmf"]
        });

        systemProps = new TestProperties(TEST_ENVIRONMENT.systemTestProperties);
        defaultSystem = systemProps.getDefaultSystem();

        account = defaultSystem.tso.account;
        jclMember = TEST_ENVIRONMENT.systemTestProperties.zosjobs.iefbr14Member;
        psJclDataSet = TEST_ENVIRONMENT.systemTestProperties.zosjobs.iefbr14PSDataSet;
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    describe("help", () => {
        it("should not change", () => {
            const response = runCliScript(__dirname + "/__scripts__/submit_data_set_help.sh", TEST_ENVIRONMENT);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toMatchSnapshot();
        });
    });


    describe("Live system tests", () => {
        it("should submit a job in an existing valid data set from a PDS member ", async () => {
            const response = runCliScript(__dirname + "/__scripts__/submit_valid_data_set.sh",
                TEST_ENVIRONMENT, [jclMember]);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toContain("jobname");
            expect(response.stdout.toString()).toContain("jobid");
        });

        it("should submit a job in an existing valid data set from a physical sequential data set", async () => {
            const response = runCliScript(__dirname + "/__scripts__/submit_valid_data_set_no_member.sh",
                TEST_ENVIRONMENT, [psJclDataSet]);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toContain("jobname");
            expect(response.stdout.toString()).toContain("jobid");
        });

        it("should submit a job in an existing valid data set (using aliases)", async () => {
            const response = runCliScript(__dirname + "/__scripts__/submit_valid_data_set_alias.sh",
                TEST_ENVIRONMENT, [jclMember]);

            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toContain("jobname");
            expect(response.stdout.toString()).toContain("jobid");
        });

        it("should fail if the data set does not exist ", async () => {
            const response = runCliScript(__dirname + "/__scripts__/submit_invalid_data_set.sh",
                TEST_ENVIRONMENT);
            expect(response.status).toBe(1);
            expect(response.stderr.toString().toLowerCase()).toContain("data");
            expect(response.stderr.toString().toLowerCase()).toContain("set");
            expect(response.stderr.toString().toLowerCase()).toContain("found");
            expect(response.stderr.toString().toLowerCase()).toContain("does.not.exist");
        });
    });

    describe("syntax errors", () => {
        it("should occur if the data set name is missing", async () => {
            const response = runCliScript(__dirname + "/__scripts__/submit_syntax_missing_data_set.sh", TEST_ENVIRONMENT);
            expect(response.status).toBe(1);
            expect(response.stdout.toString()).toBe("");
            expect(response.stderr.toString()).toMatchSnapshot();
        });

        it("should occur if an extra unknown option is specified", async () => {
            const response = runCliScript(__dirname + "/__scripts__/submit_syntax_invalid_parm.sh", TEST_ENVIRONMENT);
            expect(response.status).toBe(1);
            expect(response.stdout.toString()).toBe("");
            expect(response.stderr.toString()).toMatchSnapshot();
        });
    });
});
