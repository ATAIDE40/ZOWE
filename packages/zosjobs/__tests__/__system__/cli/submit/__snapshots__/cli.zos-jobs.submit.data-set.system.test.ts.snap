// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zos-jobs submit data-set command help should not change 1`] = `
"
 COMMAND NAME
 ------------

   data-set | ds

 DESCRIPTION
 -----------

   Submit a job (JCL) contained in a data set. The data set may be of type physical
   sequential or a PDS member. The command does not pre-validate the data set name.
   The command presents errors verbatim from the z/OSMF Jobs REST endpoints. For
   more information about z/OSMF Jobs API errors, see the z/OSMF Jobs API REST
   documentation.

 USAGE
 -----

   bright zos-jobs submit data-set <dataset> [options]

 POSITIONAL OPTIONS
 ------------------

   dataset		 (string)

      The z/OS data set containing the JCL to submit. You can specify a physical
      sequential data set (for example, \\"DATA.SET\\") or a partitioned data set
      qualified by a member (for example, \\"DATA.SET(MEMBER)\\"). The data set must be
      cataloged.

 PROFILE OPTIONS
 ---------------

   --zosmf-profile  | --zosmf-p (string) (optional)

      The name of a (zosmf) profile to load for this command execution.

 RESPONSE FORMAT OPTIONS
 -----------------------

   --response-format-filter  | --rff (array) (optional)

      Filter (include) fields in the response. Accepts an array of field/property
      names to include in the output response. You can filter JSON objects properties
      OR table columns/fields. In addition, you can use this option in conjunction
      with '--response-format-type' to reduce the output of a command to a single
      field/property or a list of a single field/property.

   --response-format-type  | --rft (string) (optional)

      The command response output format type. Must be one of the following:

      table: Formats output data as a table. Use this option when the output data is
      an array of homogeneous JSON objects. Each property of the object will become a
      column in the table.

      list: Formats output data as a list of strings. Can be used on any data type
      (JSON objects/arrays) are stringified and a new line is added after each entry
      in an array.

      object: Formats output data as a list of prettified objects (or single object).
      Can be used in place of \\"table\\" to change from tabular output to a list of
      prettified objects.

      string: Formats output data as a string. JSON objects/arrays are stringified.

      Allowed values: table, list, object, string

   --response-format-header  | --rfh (flag) (optional)

      If \\"--response-format-type table\\" is specified, include the column headers in
      the output.

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (flag) (optional)

      Produce the command response as a JSON document

   --help  | -h (flag) (optional)

      Display help text


 EXAMPLES
 --------

   Submit the JCL in the data set \\"ibmuser.cntl(deploy)\\":

	 bright zos-jobs submit data-set \\"ibmuser.cntl(deploy)\\"

{
  \\"success\\": true,
  \\"message\\": \\"The help was constructed for command: data-set.\\",
  \\"stdout\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   data-set | ds\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Submit a job (JCL) contained in a data set. The data set may be of type physical\\\\n   sequential or a PDS member. The command does not pre-validate the data set name.\\\\n   The command presents errors verbatim from the z/OSMF Jobs REST endpoints. For\\\\n   more information about z/OSMF Jobs API errors, see the z/OSMF Jobs API REST\\\\n   documentation.\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   bright zos-jobs submit data-set <dataset> [options]\\\\n\\\\n POSITIONAL OPTIONS\\\\n ------------------\\\\n\\\\n   dataset\\\\t\\\\t (string)\\\\n\\\\n      The z/OS data set containing the JCL to submit. You can specify a physical\\\\n      sequential data set (for example, \\\\\\"DATA.SET\\\\\\") or a partitioned data set\\\\n      qualified by a member (for example, \\\\\\"DATA.SET(MEMBER)\\\\\\"). The data set must be\\\\n      cataloged.\\\\n\\\\n PROFILE OPTIONS\\\\n ---------------\\\\n\\\\n   --zosmf-profile  | --zosmf-p (string) (optional)\\\\n\\\\n      The name of a (zosmf) profile to load for this command execution.\\\\n\\\\n RESPONSE FORMAT OPTIONS\\\\n -----------------------\\\\n\\\\n   --response-format-filter  | --rff (array) (optional)\\\\n\\\\n      Filter (include) fields in the response. Accepts an array of field/property\\\\n      names to include in the output response. You can filter JSON objects properties\\\\n      OR table columns/fields. In addition, you can use this option in conjunction\\\\n      with '--response-format-type' to reduce the output of a command to a single\\\\n      field/property or a list of a single field/property.\\\\n\\\\n   --response-format-type  | --rft (string) (optional)\\\\n\\\\n      The command response output format type. Must be one of the following:\\\\n\\\\n      table: Formats output data as a table. Use this option when the output data is\\\\n      an array of homogeneous JSON objects. Each property of the object will become a\\\\n      column in the table.\\\\n\\\\n      list: Formats output data as a list of strings. Can be used on any data type\\\\n      (JSON objects/arrays) are stringified and a new line is added after each entry\\\\n      in an array.\\\\n\\\\n      object: Formats output data as a list of prettified objects (or single object).\\\\n      Can be used in place of \\\\\\"table\\\\\\" to change from tabular output to a list of\\\\n      prettified objects.\\\\n\\\\n      string: Formats output data as a string. JSON objects/arrays are stringified.\\\\n\\\\n      Allowed values: table, list, object, string\\\\n\\\\n   --response-format-header  | --rfh (flag) (optional)\\\\n\\\\n      If \\\\\\"--response-format-type table\\\\\\" is specified, include the column headers in\\\\n      the output.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (flag) (optional)\\\\n\\\\n      Produce the command response as a JSON document\\\\n\\\\n   --help  | -h (flag) (optional)\\\\n\\\\n      Display help text\\\\n\\\\n\\\\n EXAMPLES\\\\n --------\\\\n\\\\n   Submit the JCL in the data set \\\\\\"ibmuser.cntl(deploy)\\\\\\":\\\\n\\\\n\\\\t bright zos-jobs submit data-set \\\\\\"ibmuser.cntl(deploy)\\\\\\"\\\\n\\\\n\\",
  \\"stderr\\": \\"\\",
  \\"data\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   data-set | ds\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Submit a job (JCL) contained in a data set. The data set may be of type physical\\\\n   sequential or a PDS member. The command does not pre-validate the data set name.\\\\n   The command presents errors verbatim from the z/OSMF Jobs REST endpoints. For\\\\n   more information about z/OSMF Jobs API errors, see the z/OSMF Jobs API REST\\\\n   documentation.\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   bright zos-jobs submit data-set <dataset> [options]\\\\n\\\\n POSITIONAL OPTIONS\\\\n ------------------\\\\n\\\\n   dataset\\\\t\\\\t (string)\\\\n\\\\n      The z/OS data set containing the JCL to submit. You can specify a physical\\\\n      sequential data set (for example, \\\\\\"DATA.SET\\\\\\") or a partitioned data set\\\\n      qualified by a member (for example, \\\\\\"DATA.SET(MEMBER)\\\\\\"). The data set must be\\\\n      cataloged.\\\\n\\\\n PROFILE OPTIONS\\\\n ---------------\\\\n\\\\n   --zosmf-profile  | --zosmf-p (string) (optional)\\\\n\\\\n      The name of a (zosmf) profile to load for this command execution.\\\\n\\\\n RESPONSE FORMAT OPTIONS\\\\n -----------------------\\\\n\\\\n   --response-format-filter  | --rff (array) (optional)\\\\n\\\\n      Filter (include) fields in the response. Accepts an array of field/property\\\\n      names to include in the output response. You can filter JSON objects properties\\\\n      OR table columns/fields. In addition, you can use this option in conjunction\\\\n      with '--response-format-type' to reduce the output of a command to a single\\\\n      field/property or a list of a single field/property.\\\\n\\\\n   --response-format-type  | --rft (string) (optional)\\\\n\\\\n      The command response output format type. Must be one of the following:\\\\n\\\\n      table: Formats output data as a table. Use this option when the output data is\\\\n      an array of homogeneous JSON objects. Each property of the object will become a\\\\n      column in the table.\\\\n\\\\n      list: Formats output data as a list of strings. Can be used on any data type\\\\n      (JSON objects/arrays) are stringified and a new line is added after each entry\\\\n      in an array.\\\\n\\\\n      object: Formats output data as a list of prettified objects (or single object).\\\\n      Can be used in place of \\\\\\"table\\\\\\" to change from tabular output to a list of\\\\n      prettified objects.\\\\n\\\\n      string: Formats output data as a string. JSON objects/arrays are stringified.\\\\n\\\\n      Allowed values: table, list, object, string\\\\n\\\\n   --response-format-header  | --rfh (flag) (optional)\\\\n\\\\n      If \\\\\\"--response-format-type table\\\\\\" is specified, include the column headers in\\\\n      the output.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (flag) (optional)\\\\n\\\\n      Produce the command response as a JSON document\\\\n\\\\n   --help  | -h (flag) (optional)\\\\n\\\\n      Display help text\\\\n\\\\n\\\\n EXAMPLES\\\\n --------\\\\n\\\\n   Submit the JCL in the data set \\\\\\"ibmuser.cntl(deploy)\\\\\\":\\\\n\\\\n\\\\t bright zos-jobs submit data-set \\\\\\"ibmuser.cntl(deploy)\\\\\\"\\\\n\\\\n\\"
}"
`;

exports[`zos-jobs submit data-set command syntax errors should occur if an extra unknown option is specified 1`] = `
"
Syntax Error:
You specified the following unknown values: \\"blah\\".

 Could not interpret them as a group, command name, or positional option.

Use \\"bright zos-jobs submit data-set blah --help\\" to view command description, usage, and options.
"
`;

exports[`zos-jobs submit data-set command syntax errors should occur if the data set name is missing 1`] = `
"
Syntax Error:
Missing Positional Option:
dataset

Option Description:
The z/OS data set containing the JCL to submit. You can specify a physical
sequential data set (for example, \\"DATA.SET\\") or a partitioned data set
qualified by a member (for example, \\"DATA.SET(MEMBER)\\"). The data set must be
cataloged.

Use \\"bright zos-jobs submit data-set --help\\" to view command description, usage, and options.
"
`;
